File: src/main/java/at/xirado/tuner/Application.java
/*
 * Copyright 2022 Marcel Korzonek and the Tuner contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.xirado.tuner;

import at.xirado.tuner.audio.AudioManager;
import at.xirado.tuner.config.ConfigLoader;
import at.xirado.tuner.config.TunerConfiguration;
import at.xirado.tuner.data.guild.GuildManager;
import at.xirado.tuner.interaction.InteractionManager;
import at.xirado.tuner.listener.ReadyListener;
import at.xirado.tuner.listener.SlashCommandListener;
import ch.qos.logback.classic.Level;
import com.sedmelluq.discord.lavaplayer.jdaudp.NativeAudioSendFactory;
import net.dv8tion.jda.api.GatewayEncoding;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder;
import net.dv8tion.jda.api.sharding.ShardManager;
import net.dv8tion.jda.api.utils.cache.CacheFlag;
import okhttp3.OkHttpClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.security.auth.login.LoginException;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;

public class Application {

    private static Logger LOG = LoggerFactory.getLogger(Application.class);
    private static Application instance;

    public static void main(String[] args) throws IOException, LoginException, ClassNotFoundException {
        List<String> argsList = Arrays.asList(args);
        if (!argsList.contains("--noclear")) {
            System.out.print("\033[2J\033[H"); // Clears the terminal
        }

        if (argsList.contains("--debug")) {
            LOG.info("Started with \"--debug\" argument.");
            ((ch.qos.logback.classic.Logger) LoggerFactory.getLogger("ROOT")).setLevel(Level.DEBUG);
        }

        Thread.currentThread().setName("Tuner Main-Thread");
        new Application();
    }

    public static Application getApplication() {
        return instance;
    }

    private final OkHttpClient httpClient = new OkHttpClient();
    private final TunerConfiguration tunerConfiguration;
    private final ShardManager shardManager;
    private final InteractionManager interactionManager;
    private final AudioManager audioManager;
    private final GuildManager guildManager;

    private Application() throws IOException, LoginException, ClassNotFoundException {
        instance = this;
        this.tunerConfiguration = new TunerConfiguration(ConfigLoader.loadFileAsYaml("config.yml", true));

        if (tunerConfiguration.getDiscordToken() == null || tunerConfiguration.getDiscordToken().isEmpty()) {
            throw new IllegalArgumentException("config.yml does not contain \"discord_token\" property!");
        }

        String token = tunerConfiguration.getDiscordToken();

        this.shardManager = DefaultShardManagerBuilder.createDefault(token, GatewayIntent.GUILD_MESSAGES, GatewayIntent.GUILD_VOICE_STATES)
                .setBulkDeleteSplittingEnabled(false)
                .setGatewayEncoding(GatewayEncoding.ETF)
                .setActivity(Activity.listening("music"))
                .disableCache(CacheFlag.EMOTE)
                .setAudioSendFactory(new NativeAudioSendFactory())
                .addEventListeners(new ReadyListener(this), new SlashCommandListener(this))
                .build();

        this.interactionManager = new InteractionManager(this);
        this.audioManager = new AudioManager(this);
        Class.forName("at.xirado.tuner.db.Database");
        this.guildManager = new GuildManager(shardManager);

    }

    public TunerConfiguration getTunerConfiguration() {
        return tunerConfiguration;
    }

    public ShardManager getShardManager() {
        return shardManager;
    }

    public InteractionManager getInteractionManager() {
        return interactionManager;
    }

    public AudioManager getAudioManager() {
        return audioManager;
    }

    public OkHttpClient getHttpClient() {
        return httpClient;
    }

    public GuildManager getGuildManager() {
        return guildManager;
    }
}

File: src/main/java/at/xirado/tuner/Application.kt
/*
 * Copyright 2022 Marcel Korzonek and the Tuner contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:JvmName("Main")
package at.xirado.tuner

import at.xirado.tuner.Application.Companion.application
import at.xirado.tuner.audio.AudioManager
import at.xirado.tuner.config.ConfigLoader
import at.xirado.tuner.config.TunerConfiguration
import at.xirado.tuner.data.GuildManager
import at.xirado.tuner.interaction.InteractionHandler
import at.xirado.tuner.interaction.MultiBotManager
import at.xirado.tuner.listener.InteractionListener
import at.xirado.tuner.listener.RegistrationListener
import at.xirado.tuner.listener.VoiceListener
import at.xirado.tuner.log.DiscordWebhookAppender
import at.xirado.tuner.util.Util
import ch.qos.logback.classic.Level
import com.sedmelluq.discord.lavaplayer.jdaudp.NativeAudioSendFactory
import dev.minn.jda.ktx.getDefaultScope
import net.dv8tion.jda.api.GatewayEncoding
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.cache.CacheFlag
import okhttp3.OkHttpClient
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import java.nio.charset.StandardCharsets
import java.util.Base64
import java.util.concurrent.ConcurrentHashMap

class Application {

    companion object {
        val log = LoggerFactory.getLogger(Application::class.java) as Logger
        lateinit var application: Application
    }

    val tunerConfig: TunerConfiguration
    val coroutineScope = getDefaultScope()
    val bots = mutableMapOf<Long, JDA>()
    val audioManagers = mutableMapOf<Long, AudioManager>()
    val httpClient = OkHttpClient()
    val guildManager: GuildManager
    val multiBotManager: MultiBotManager
    val interactionHandler: InteractionHandler

    init {
        application = this
        tunerConfig = TunerConfiguration(ConfigLoader.loadFileAsYaml("config.yml", true))
        multiBotManager = MultiBotManager(this)
        interactionHandler = InteractionHandler(this)
        DiscordWebhookAppender.init(this)
        if (tunerConfig.discordTokens.isEmpty())
            throw IllegalArgumentException("\"discord_tokens\" property does not exist or is empty!")

        val tokens = tunerConfig.discordTokens

        tokens.forEachIndexed { index, token ->
            val contextMap = ConcurrentHashMap<String, String>()
            contextMap["jda.shard"] = "Shard $index"

            val userId = getUserIdFromToken(token)
            audioManagers[userId] = AudioManager(this, userId)
            bots[userId] = JDABuilder.createDefault(token, GatewayIntent.GUILD_MESSAGES, GatewayIntent.GUILD_VOICE_STATES, GatewayIntent.GUILD_MESSAGES)
                .setBulkDeleteSplittingEnabled(false)
                .setGatewayEncoding(GatewayEncoding.ETF)
                .setActivity(Activity.listening("music (Shard ${index+1})"))
                .disableCache(CacheFlag.EMOTE)
                .setContextMap(contextMap)
                .setContextEnabled(true)
                .setAudioSendFactory(NativeAudioSendFactory())
                .addEventListeners(InteractionListener(this), RegistrationListener(this), VoiceListener(this))
                .build()
        }
        Class.forName("at.xirado.tuner.data.Database")
        guildManager = GuildManager(this)
    }
}

fun main(args: Array<String>) {
    if ("--noclear" !in args)
        Util.clearScreen()

    if ("--debug" in args)
        (LoggerFactory.getLogger("ROOT") as ch.qos.logback.classic.Logger).level = Level.DEBUG

    Thread.setDefaultUncaughtExceptionHandler { _, e -> Application.log.error("An unhandled exception was encountered", e)}
    Thread.currentThread().name = "Tuner Main-Thread"
    Application()
}

private fun getUserIdFromToken(token: String): Long {
    val userIdEncoded = token.split(".")[0]
    return String(Base64.getDecoder().decode(userIdEncoded), StandardCharsets.UTF_8).toLong()
}